Remove __start__ stack, log backward compile to empty stack

jamesjwu opened the pull request:
Summary:
This diff removes "__start__" from all stacks in Pt2 Compile Events, as it's unnecessary. 

It also starts logging events for backward compile, because otherwise we have no toplevel event representing full backward compilation. This gives us a toplevel event outside of the inductor compile.

Test Plan:
New chromium events: 

https://interncache-all.fbcdn.net/manifold/perfetto-artifacts/tree/ui/index.html?url=https%3A%2F%2Finterncache-all.fbcdn.net%2Fmanifold%2Ftlparse_reports%2Ftree%2Flogs%2Fjjwu%2Fcustom%2Fstuff4%2Fchromium_events.json#!/viewer?url=https%3A%2F%2Finterncache-all.fbcdn.net%2Fmanifold%2Ftlparse_reports%2Ftree%2Flogs%2Fjjwu%2Fcustom%2Fstuff4%2Fchromium_events.json&local_cache_key

New tlparse:
https://interncache-all.fbcdn.net/manifold/tlparse_reports/tree/logs/jjwu/custom/stuff4/index.html

New scuba icicle view, still good: https://fburl.com/scuba/pt2_compile_events/z6gr3z53

Differential Revision: D65832045




cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang @aakhundov

masnesral commented on torch/_inductor/compile_fx.py:
```diff
@@ -1632,7 +1632,11 @@ def partition_fn(
             model: GraphModule, example_inputs: List[InputType]
         ) -> Union[CompiledFxGraph, str]:
-            with dynamo_utils.dynamo_timed("compile_fx.<locals>.bw_compiler"):
+            with dynamo_utils.dynamo_timed(
```
I think I also tried to get the full backward time for purposes of logging to dynamo_compile, but I put it here instead: https://github.com/pytorch/pytorch/blob/267641f6f1c011dba7fd5e514d766598035e06dd/torch/_functorch/_aot_autograd/runtime_wrappers.py#L2011

I guess we should pick one?

jamesjwu replied:
Oh if you have this already I don't need to land mine

jamesjwu replied:
Yeah, I'm just going to not add it here, and instead have you add it to wherever you had entire_backward_compile.

ezyang commented:
deferring this review to @masnesral

jamesjwu committed:
[PT2 Compile Events] Improvements to `stack` logging (#140431)

Summary:

This diff implements multiple improvements to the `stack` field in PT2 Compile Events. 

- First, it removes `__start__` from all stacks in Pt2 Compile Events, as it's unnecessary. 

Finally, it fixes a bug where we are logging the entire chromium stack instead of just the stack involving events that are logged to scuba. At first, I thought it was safe to log the entire stack, but it leads to icicle views that "lie" about time taken for intermediate events that are logged to TLParse, but not Scuba. 

As an example, https://fburl.com/scuba/pt2_compile_events/11pqmcnt

this icicle view shows that AOTAutogradCache.load is only 50% of backend_compile. But in reality, since AOTAutogradCache.load doesn't have its own event in the table, it's only showing the portions of **events whose stacks include AOTAutogradCache.load**, which is incorrect. The actual duration of AOTAutogradCache.load is actually the entire backend_compile time. {F1958436460}

In reality, the stack in TLParse looks more like this:
 {F1958437327} 

Where the extra time is accounted for inside of AOTAutogradCache.load and its sub events, but unaccounted for by the aggregated stack. 

Therefore, we should only log events to the Scuba stack that are themselves in the scuba dataset to begin with. To do that, we keep a pt2_stack along with a regular stack, which is the subset of the regular stack that gets logged to scuba.

This maintains the ability to register event metadata and the regular stack in TLParses.

Test Plan:
Chromium events unchanged: 
https://interncache-all.fbcdn.net/manifold/perfetto-artifacts/tree/ui/index.html?url=https%3A%2F%2Finterncache-all.fbcdn.net%2Fmanifold%2Ftlparse_reports%2Ftree%2Flogs%2Fjjwu%2Fcustom%2FD65832045%2Fchromium_events.json#!/viewer?url=https%3A%2F%2Finterncache-all.fbcdn.net%2Fmanifold%2Ftlparse_reports%2Ftree%2Flogs%2Fjjwu%2Fcustom%2FD65832045%2Fchromium_events.json&local_cache_key

Tlparse unchanged:

https://interncache-all.fbcdn.net/manifold/tlparse_reports/tree/logs/jjwu/custom/D65832045/index.html


New scuba icicle view, much simpler and more accurate: 
https://fburl.com/scuba/pt2_compile_events/3ian8gnq

 {F1958439008}

Reviewed By: ezyang, masnesral

Differential Revision: D65832045
